Add a material interface reaction

parmar0 opened the pull request:
Add a material interface reaction. Closes #27736

## Reason
Need to model variable reaction rates.

## Design
Adding a class with reaction rate coefficients specified by AD Material Properties.

## Impact
Allows users to model variable reaction rates.

parmar0 committed:
Add a material interface reaction. Closes #27736

ke7kto commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_steady.i:
```diff
@@ -0,0 +1,160 @@
+    type = ADMatInterfaceReaction
+    variable = u
+    neighbor_var = 'v'
+    boundary = 'right'
```
This boundary should be `primary0_interface`

ke7kto commented on test/tests/interfacekernels/1d_interface/tests:
```diff
@@ -124,7 +124,14 @@
+    type = CSVDiff
+    input = 'ADMatreaction_1D_steady.i'
+    csvdiff = 'reaction_1D_steady_out.csv'
+    issues = '#13365'
```
wrong issue number here. Line 118 should be #13365 and 131 should be #27736

ke7kto reviewed (changes requested):
Good first draft, but I think I see some errors.

parmar0 committed:
applied suggestions from code review

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2256222/) on 34794f0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27737/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27737/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4409da00956c2511d3f6c6db886956a265cd4dd9`

parmar0 committed:
code with suggested style changes

code with deleted whitespace

addressing errors related to CSV files from review

standardized test output files

adding unique requirement in tests

revised tests to resolve errors

lindsayad commented:
You're failing precheck because of trailing whitespace
```
##########################################################################
ERROR: The following files contain trailing whitespace after applying your patch:
	test/tests/interfacekernels/1d_interface/ADMatreaction_1D_transient.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.

ERROR: The following files do not contain a newline character before EOF:
	test/tests/interfacekernels/1d_interface/ADMatreaction_1D_transient.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
```

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2259027/) on de3fe1a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27737/site)

This comment will be updated on new commits.

ke7kto commented:
@lindsayad I noticed that Akanksha's csvdiff was failing because she wasn't specifying the right csv output file in her `test/tests/interfacekernels/1d_interface/ADMatreaction_1D_steady.i`. She's put together a fix for that. I had initially thought that the best practice would be to use the same gold file and save space, but is there a chance of a race condition here if we go that route? Should we instead create a new gold csv output - there should be no difference between the `reaction_1D_steady` and `ADMatReaction_1D_steady` files.

lindsayad commented:
I usually create a symlink in the gold file directory. This removes the possibility of a race condition

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2261605/) on f844c4f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>125972</th>
      <th colspan=4>#27737 f844c4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.09%</td>
      <td>85.09%</td>
      <td>+0.00%</td>
      <td>90.91%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104026</td>
      <td>104046</td>
      <td>+20</td>
      <td>20</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18226</td>
      <td>18228</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27737/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27737/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27737/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad reviewed (changes requested):
Please remove the merge commit. You can use rebase if you need something new on next

parmar0 committed:
Revert changes to tex files

ke7kto commented:
@lindsayad I think the tests should all be good now, but we have a couple of the modules test that are timing out.

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+With the first order reaction rate assuming a quasi-steady-state
+
+\begin{equation}
+Reaction Rate = \frac {\partial C_1} {\partial t} = k_f C_1 - k_b C_2 \approx 0
```
```suggestion
\textrm{Reaction Rate} = \frac {\partial C_1} {\partial t} = k_f C_1 - k_b C_2 \approx 0
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+Reaction Rate = \frac {\partial C_1} {\partial t} = k_f C_1 - k_b C_2 \approx 0
+\end{equation}
+
+where $C_1$ is the specie concentration in domain 1, $C_2$ is the specie concentration in domain 2, $k_f$ is the forward reaction coefficient, and $k_b$ is the backward reaction coefficient. `ADMatInterfaceReaction` object is used to impose this condition. Associated kernel is:
```
```suggestion
where $C_1$ is the specie concentration in domain 1, $C_2$ is the specie concentration in domain 2, $k_f$ is the forward reaction coefficient, and $k_b$ is the backward reaction coefficient. `ADMatInterfaceReaction` object is used to impose this condition.
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+
+where $C_1$ is the specie concentration in domain 1, $C_2$ is the specie concentration in domain 2, $k_f$ is the forward reaction coefficient, and $k_b$ is the backward reaction coefficient. `ADMatInterfaceReaction` object is used to impose this condition. Associated kernel is:
+
+!syntax description /InterfaceKernels/ADMatInterfaceReaction
+
+## Overview
+
+!! Replace these lines with information regarding the ADMatInterfaceReaction object.
+
```
```suggestion
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+
+!! Replace these lines with information regarding the ADMatInterfaceReaction object.
+
+## Example Input File Syntax
+
+!! Describe and include an example of how to use the ADMatInterfaceReaction object.
+
```
```suggestion
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+!syntax children /InterfaceKernels/ADMatInterfaceReaction
+
+
+[/ADMatInterfaceReaction.C]
+
+[/ADMatInterfaceReaction.h]
+
```
```suggestion
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+
+[/ADMatInterfaceReaction.h]
+
+In addition, fluxes are matched from both domains, this could be achieved by  [`InterfaceDiffusion`](/InterfaceKernels/index.md). 
+
+Both kernels at the interface work together to give full mathematical and physical meaning of the problem.
```
This content should be moved up with the other description text

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+
+Both kernels at the interface work together to give full mathematical and physical meaning of the problem.
+
+Two examples (steady-state and transient-state) are shown in the MOOSE test directory, 
+
+[1d_interface/reaction_1D_steady.i]
+
+[1d_interface/reaction_1D_transient.i]
+
+
```
```suggestion
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,54 @@
+[1d_interface/reaction_1D_transient.i]
+
+
+## Example Input Syntax
+
+!listing test/tests/interfacekernels/1d_interface/ADMatreaction_1D_steady.i start=[./interface_reaction] end=[../] include-end=true
+!listing test/tests/interfacekernels/1d_interface/ADMatreaction_1D_transient.i start=[./interface_reaction] end=[../] include-end=true
```
If you want to include this, then move it before `!syntax inputs` and walk the user through the blocks

lindsayad commented on framework/include/interfacekernels/ADMatInterfaceReaction.h:
```diff
@@ -0,0 +1,35 @@
+ * Implements a reaction to establish ReactionRate=k_f*u-k_b*v
+ * at interface.
+ */
+
```
```suggestion
```

lindsayad commented on framework/include/interfacekernels/ADMatInterfaceReaction.h:
```diff
@@ -0,0 +1,35 @@
+ * at interface.
+ */
+
+class ADMatInterfaceReaction : public ADInterfaceKernel ///ADInterfaceKernel ??
```
```suggestion
class ADMatInterfaceReaction : public ADInterfaceKernel
```

lindsayad commented on framework/include/interfacekernels/ADMatInterfaceReaction.h:
```diff
@@ -0,0 +1,35 @@
+
+protected:
+  virtual ADReal computeQpResidual(Moose::DGResidualType type) override;
+  // virtual ADReal computeQpJacobian(Moose::DGJacobianType type) override;
```
```suggestion
```

lindsayad commented on framework/src/interfacekernels/ADMatInterfaceReaction.C:
```diff
@@ -0,0 +1,54 @@
+ADMatInterfaceReaction::ADMatInterfaceReaction(const InputParameters & parameters)
+  : ADInterfaceKernel(parameters),
+    _kf(getADMaterialProperty<Real>("forward_rate")),
+    _kb(getADMaterialProperty<Real>("backward_rate"))
```
```suggestion
    _kb(getNeighborADMaterialProperty<Real>("backward_rate"))
```

lindsayad commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_steady.i:
```diff
@@ -0,0 +1,172 @@
+#
+# At the interface consider the following
+#
+# (a) Fluxes are matched from both domains (InterfaceDiffusion kernel)
+#
+# (b) First-order reaction is R = kf*u - kb*v
```
Are you sure the fluxes are matched? The condition you are implicitly setting at the interface for the u variable is this: $-D_u \nabla u \cdot \hat{n} = -D_v \nabla v \cdot \hat{n} + k_f u - k_b v$, so I hope that the fluxes do not match

lindsayad replied:
Can you comment on this?

ke7kto replied:
I agree with your assessment of the implicit condition at the interface. It looks like a Robin BC on both sides to me. 

The flux at both sides of the interface should be matched, with some specified offset, which may change and may also be linked to a material property. The - and + side fluxes should negate each other and the u-v value terms should also negate, though I suppose there might be other potential values of the solution if the flux/value specifications conflict with each other. 

Clearly I'm missing something in what you're trying to tell me, but I'd also point out that this is verbatim copied from the `reaction_1D_steady.i` test, so whatever the resolution is, we'll need to change it there too.

lindsayad replied:
I think we should delete these lines and replace them by the interface condition we are weakly enforcing which is $-D_u \nabla u \cdot \hat{n} = -D_v \nabla v \cdot \hat{n} + k_f u - k_b v$ Basically we'll want to match the equation we put into the documentation

ke7kto replied:
I'll write the weakly enforced boundary condition we're applying, but I really want to be explicit about exactly how we're applying that condition, so I'd recommend not deleting the explanation of the constituent parts. The class this addition is based off of was my first class I really dug into in the moose framework, and it took me way too long to understand what each piece was doing, and what I needed to put in to make sure the physics was right. Of course, the documentation was much less robust back then.

lindsayad replied:
I had to throw some `SideDiffusiveFluxIntegral` postprocessors on the non-AD test locally to test this out because I did not believe it. But indeed the diffusive fluxes are equivalent like you said! So instead of simply $-D_u \nabla u \cdot \hat{n} = -D_v \nabla v \cdot \hat{n} + k_f u - k_b v$ we have $-D_u \nabla u\cdot\hat{n} = -D_v \nabla v \cdot \hat{n}$ and $k_f u = k_b v$. It's not entirely clear in my head what is constraining the fluxes to be the same on either side of the interface but my feeling is the requirement for a constant gradient on either subdomain.

But yea, I guess we can keep this comment!

lindsayad replied:
One thing you might consider is writing: `R = kf*u - kb*v = 0`

lindsayad commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_steady.i:
```diff
@@ -0,0 +1,172 @@
+    nx = 10
+    xmax = 2
+  []
+  [./subdomain1]
+    input = gen
+    type = SubdomainBoundingBoxGenerator
+    bottom_left = '1.0 0 0'
+    block_id = 1
+    top_right = '2.0 1.0 0'
+  [../]
```
Remove all `./` and `../` syntax from the input as they're no longer needed by the parser

lindsayad commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_steady.i:
```diff
@@ -0,0 +1,172 @@
+  [../]
+[]
+
+[Preconditioning]
+  [./smp]
+    type = SMP
+    full = true
+  [../]
+[]
+
+[Executioner]
+  type = Steady
+  solve_type = PJFNK
+  nl_rel_tol = 1e-10
+[]
```
```suggestion
[Executioner]
  type = Steady
  solve_type = NEWTON
  nl_rel_tol = 1e-10
[]
```

lindsayad commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_transient.i:
```diff
@@ -0,0 +1,157 @@
+      nx = 10
+      xmax = 2
+    []
+    [./subdomain1]
+      input = gen
+      type = SubdomainBoundingBoxGenerator
+      bottom_left = '1.0 0 0'
+      block_id = 1
+      top_right = '2.0 1.0 0'
+    [../]
```
as for other input, remove `./` and `../` syntax

lindsayad commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_transient.i:
```diff
@@ -0,0 +1,157 @@
+    [../]
+  []
+
+  [Preconditioning]
+    [./smp]
+      type = SMP
+      full = true
+    [../]
+  []
```
this block is not needed for Newton as it's added automatically but you can keep it if you want

lindsayad commented on test/tests/interfacekernels/1d_interface/ADMatreaction_1D_transient.i:
```diff
@@ -0,0 +1,157 @@
+    show_var_residual_norms = true
+  []
+
+ # [Postprocessors]
+ #    [./elemental_error_u]
+ #      type = ElementL2Error
+#       function = -0.2*x+1
+#       variable = 'u'
+ #      block = '0'
+ #    [../]
+#     [./elemental_error_v]
+ #      type = ElementL2Error
+ #      function = -0.4*x+0.8
+ #      variable = 'v'
+#       block = '1'
+ #    [../]
+ #  []
```
Remove if not used

lindsayad commented on test/tests/interfacekernels/1d_interface/tests:
```diff
@@ -124,7 +124,22 @@
+    input = 'ADMatreaction_1D_transient.i'
+    exodiff = 'ADMatreaction_1D_transient_out.e'
+    issues = '#27736'
+    requirement = "The result of an interface reaction kernel computed using automatic differentiation shall match the analytic solution in a transient simulation."
```
are you testing against an analytic solution in this transient test?

lindsayad commented:
thanks for your contribution @parmar0 !

parmar0 committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

parmar0 committed:
Applied suggestions for .md and tests files

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 2fe4657bc59b93eb8179d22fa1b3ba137361b6b3
+Subproject commit 79f8bcf4c31721323622414ef59d33f3db7ac7c5
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

ke7kto commented:
@lindsayad We've tried to remove any references to the large_media submodule (revert to previous commit, force push), but it seems that we're still trying to change it somehow. We haven't touched that folder. Is this something we need to put more time into, or is there an easy way to remove that from the PR at this point

cticenhour commented:
@ke7kto The easiest way without a rebase at this point, with the current history, is fixing up the commit that contains the update.

```
cd ~/projects/moose/large_media   # or wherever this MOOSE is
git checkout 2fe4657bc59b93eb8179d22fa1b3ba137361b6b3
cd ..
git add large_media
git commit --fixup 9463948f019e99c8ac39f6613a0135f156e6c1f9
git rebase -i bac2b46800b29699032680442b035c2266eaa99a --autosquash
```

cticenhour commented:
@ke7kto Looks like your last formatting commit didn't apply through the push - did you not see it in the rebase?

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,31 @@
+
+where $C_1$ is the specie concentration in domain 1, $C_2$ is the specie concentration in domain 2, $k_f$ is the forward reaction coefficient, and $k_b$ is the backward reaction coefficient. `ADMatInterfaceReaction` object is used to impose this condition.
+
+[`InterfaceDiffusion`](/InterfaceKernels/index.md) is also used in this case to control flux at  the interface. However, the flux is not well-defined at boundaries. To address this issue, InterfaceReaction kernel applies a condition to constrain the discontinuity across the interface.
```
```suggestion
[InterfaceDiffusion.md] is also used in this case to control flux at  the interface. However, the flux is not well-defined at boundaries. To address this issue, InterfaceReaction kernel applies a condition to constrain the discontinuity across the interface.
```

lindsayad commented on framework/doc/content/source/interfacekernels/ADMatInterfaceReaction.md:
```diff
@@ -0,0 +1,31 @@
+
+where $C_1$ is the specie concentration in domain 1, $C_2$ is the specie concentration in domain 2, $k_f$ is the forward reaction coefficient, and $k_b$ is the backward reaction coefficient. `ADMatInterfaceReaction` object is used to impose this condition.
+
+[`InterfaceDiffusion`](/InterfaceKernels/index.md) is also used in this case to control flux at  the interface. However, the flux is not well-defined at boundaries. To address this issue, InterfaceReaction kernel applies a condition to constrain the discontinuity across the interface.
```
> However, the flux is not well-defined at boundaries. To address this issue, InterfaceReaction kernel applies a condition to constrain the discontinuity across the interface.

There is a clear definition of the flux on + and - sides of interfaces and at boundaries. Perhaps you should say "The interface reaction kernel introduces a discontinuity in flux at the interface." ?

ke7kto replied:
Would "However, the flux is not [well-defined](https://en.wikipedia.org/wiki/Well-defined_expression) across the interface (potential discontinuity)" work?

I don't think we're really introducing a discontinuity as much as we are constraining it. Unless I'm mistaken, the `InterfaceDiffusion` interfacekernel only constrains the + and - side fluxes to be equal. In the absence of error, we would maintain the initial difference across the interface, but that's not actively being constrained unless some other physics is in play.

lindsayad replied:
I think the clearest documentation would be to write an equation describing the interface conditions being applied when both `ADMatInterfaceReaction` and `InterfaceDiffusion` are applied at the same interface

lindsayad commented on test/tests/interfacekernels/1d_interface/tests:
```diff
@@ -124,7 +124,22 @@
+    input = 'ADMatreaction_1D_transient.i'
+    exodiff = 'ADMatreaction_1D_transient_out.e'
+    issues = '#27736'
+    requirement = "The result of an interface reaction kernel computed using automatic differentiation shall be consistent with the results of the non-AD transient interfaceReaction test."
```
```suggestion
    requirement = "The result of an interface reaction kernel computed using automatic differentiation shall be consistent with the results of the non-AD transient interface reaction test."
```

ke7kto committed:
Code and documentation formatting updates

ke7kto committed:
spelling correction on tests file

ke7kto committed:
Added equation representing combined InterfaceDiffusion and ADMatInterfaceReaction terms to documentation

ke7kto committed:
Updated comments in test files to reflect the effective interface condition created in the test

ke7kto committed:
Remove unicode character from test file

ke7kto committed:
adjust comments on 1d steady interface test

ke7kto committed:
Update reaction_1D_steady.i

ke7kto committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad reviewed (approved):
thanks for your hard work on this @ke7kto and opening my eyes to these neat interface conditions. Will merge pending tests passing
